test/contracts/OVM/bridge/OVM_L1MultiMessageRelayer.ts

```diff
@@ -0,0 +1,134 @@
+  let Factory__OVM_L1MultiMessageRelayer: ContractFactory
+  let Mock__OVM_L1CrossDomainMessenger: MockContract
+  let messages: any[]
+  
+  before(async () => {
+
+    AddressManager = await makeAddressManager()
+
+    Factory__OVM_L1MultiMessageRelayer = await ethers.getContractFactory(
+      'OVM_L1MultiMessageRelayer'
+    )
+
+    Mock__OVM_L1CrossDomainMessenger = await smockit(
+      await ethers.getContractFactory('OVM_L1CrossDomainMessenger')
+    )
+
+    // set the address of the mock contract to target
+    await AddressManager.setAddress(
+      'OVM_L1CrossDomainMessenger',
+      Mock__OVM_L1CrossDomainMessenger.address
+    )
+
+    // set the signer as the address required by access control
+    await AddressManager.setAddress(
+      'OVM_L2BatchMessageRelayer',
+      signer.getAddress()
+    )
+
+    // define a dummy proof to satisfy the abi
+    let dummyProof = {
+      stateRoot: NON_NULL_BYTES32,
+      stateRootBatchHeader: DUMMY_BATCH_HEADERS[0],
+      stateRootProof: DUMMY_BATCH_PROOFS[0],
+      stateTrieWitness: toHexString("some bytes"),
+      storageTrieWitness: toHexString("some more bytes")
+    } 
+
+    // create a few dummy messages to relay
+    let m1 = {
+      target: "0x1100000000000000000000000000000000000000",
+      message: NON_NULL_BYTES32,
+      sender: "0x2200000000000000000000000000000000000000",
+      messageNonce: 1,
+      proof: dummyProof,
+      calldata: "0xaabbccddeeff"
+    }
+    let m2 = {
+      target: "0x1100000000000000000000000000000000000000",
+      message: NON_NULL_BYTES32,
+      sender: "0x2200000000000000000000000000000000000000",
+      messageNonce: 2,
+      proof: dummyProof,
+      calldata: "0xaabbccddeeff"
+    }
+    let m3 = {
+      target: "0x1100000000000000000000000000000000000000",
+      message: NON_NULL_BYTES32,
+      sender: "0x2200000000000000000000000000000000000000",
+      messageNonce: 2,
+      proof: dummyProof,
+      calldata: "0xaabbccddeeff"
+    }
+    messages = [m1, m2, m3]
+  }
+)
+
+  let OVM_L1MultiMessageRelayer: Contract
+
+  beforeEach(async () => {
+    OVM_L1MultiMessageRelayer = await Factory__OVM_L1MultiMessageRelayer.deploy(
+      AddressManager.address
+    )
+
+    // set the address of the OVM_L1MultiMessageRelayer, which the OVM_L1CrossDomainMessenger will
+    // check in its onlyRelayer modifier
+    await AddressManager.setAddress(
+      'OVM_L2MessageRelayer', // This is the string currently used in the AddressManager
+      OVM_L1MultiMessageRelayer.address
+    )
+    // set the mock return value
+    Mock__OVM_L1CrossDomainMessenger.smocked.relayMessage.will.return()
+  })
+
+  describe('batchRelayMessages', () => {
+    it('Successfully relay multiple messages', async () => {
+      await OVM_L1MultiMessageRelayer.batchRelayMessages(
+        messages
+      )
+      await expect(
+        Mock__OVM_L1CrossDomainMessenger.smocked.relayMessage.calls.length
+      ).to.deep.equal(messages.length)
```

Comment:
Nit: I think we should be able to explicitly assert `(calls).to.deep.equal(messages)` here if the types work out how I'm thinking.

Reply:
It doesn't seem to be straightforward. I tried `deep.equal` and `deep.include` but neither are working. I also tried iterating  over `calls` and `messages` with no luck. 

Each message looks like this:

```
{ target: '0x1100000000000000000000000000000000000000',
    message: '0x1111111111111111111111111111111111111111111111111111111111111111',
    sender: '0x2200000000000000000000000000000000000000',
    messageNonce: 1,
    proof:
     { stateRoot: '0x1111111111111111111111111111111111111111111111111111111111111111',
       stateRootBatchHeader: [Object],
       stateRootProof: [Object],
       stateTrieWitness: '0x736f6d65206279746573',
       storageTrieWitness: '0x736f6d65206d6f7265206279746573' } }
```

Whereas the calls object looks like the attached, with mangled keys for the arguments: 
[calls.txt](https://github.com/ethereum-optimism/contracts-v2/files/5914585/calls.txt)

Can it be done, yes, and it'd be nice to have built into smock to decode the calls object, but it's not worth the effort right now IMO.